source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/ObjectModelTest.cpp

```diff
@@ -450,5 +458,596 @@ namespace AdaptiveCardsSharedModelUnitTest
             // default value test
             runWrapTest<ToggleInput>(body, 1, false);
         }
+
+        TEST_METHOD(ElementFallbackSerializationTest)
+        {
+            std::string cardStr = R"card({
+                "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+                "type" : "AdaptiveCard",
+                "version" : "1.2",
+                "body" : [
+                    {
+                        "type": "TextBlock",
+                        "text" : "TextBlock with fallback drop",
+                        "fallback" : "drop"
+                    },
+                    {
+                        "type": "TextBlock",
+                        "text" : "TextBlock with fallback content",
+                        "fallback" :
+                        {
+                            "type": "TextBlock",
+                            "text" : "fallback content goes here"
+                        }
+                    }
+                ]
+            })card";
+            auto parseResult = AdaptiveCard::DeserializeFromString(cardStr, "1.2");
+            auto card = parseResult->GetAdaptiveCard();
+            auto body = card->GetBody();
+            auto textBlock1 = std::static_pointer_cast<TextBlock>(body.at(0));
+            Assert::IsTrue(FallbackType::Drop == textBlock1->GetFallbackType(), L"Drop type");
+            auto textBlock2 = std::static_pointer_cast<TextBlock>(body.at(1));
+            Assert::IsTrue(FallbackType::Content == textBlock2->GetFallbackType(), L"Content type");
+            auto fallbackTextBlock = std::static_pointer_cast<TextBlock>(textBlock2->GetFallbackContent());
+            Assert::AreEqual(fallbackTextBlock->GetText().c_str(), "fallback content goes here");
+            auto serializedCard = card->Serialize();
+            Assert::AreEqual(serializedCard.c_str(),
+                "{\"actions\":[],\"body\":[{\"fallback\":\"drop\",\"text\":\"TextBlock with fallback drop\",\"type\":\"TextBlock\"},{\"fallback\":{\"text\":\"fallback content goes here\",\"type\":\"TextBlock\"},\"text\":\"TextBlock with fallback content\",\"type\":\"TextBlock\"}],\"type\":\"AdaptiveCard\",\"version\":\"1.2\"}\n");
+        }
+
+        TEST_METHOD(ActionFallbackSerializationTest)
+        {
+            std::string cardStr = R"card({
+                "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+                "type" : "AdaptiveCard",
+                "version" : "1.2",
+                "body" : [
+                    {
+                        "type": "TextBlock",
+                        "text" : "test text"
+                    }
+                ],
+               	"actions": [
+               		{
+               			"type": "Action.Submit",
```

Comment:
>			 [](start = 15, length = 3)

Tabs

Reply:
ugh

---
In reply to: [253127516](https://github.com/Microsoft/AdaptiveCards/pull/2351#discussion_r253127516) [](ancestors = 253127516)